### PR TITLE
Move from Argparse to Omegaconf

### PR DESCRIPTION
**Pull request by vturrisi**
Opening the PR now so that we can properly visualize the changes and eventually get the tests rolling.

### CODE REVIEW

**Comment by turian** [comment]
@vturrisi Why not use [hydra](https://hydra.cc/) from FAIR. It is a more powerful, and can easily convert to and from `OmegaConf` options?

**Comment by turian** [comment]
I have used hydra in the past and could give feedback and code review

**Comment by vturrisi** [comment]
@turian yes, using hydra is another possibility. What would be the main advantages over omegaconf?

**Comment by turian** [comment]
@vturrisi Hydra is a library that builds on Omegaconf, so it's more powerful. (Nonetheless, you can easily cast things to OmegaConf if need be.) But it's pretty easy to port to it.

TBH I am not super familiar with OmegaConf because I just just Hydra and it seems to handle the casts to OmegaConf automatically as necessary.

Here is what Hydra does that's cool, let me know if OmegaConf does all this:
* You group your configs into different yaml files in config, maybe them easy to find.
* You can compose the YAML files but easily override them.

Skim the intro: https://hydra.cc/docs/intro/
and you'll see what I mean.

There's also the [Structured Configs](https://hydra.cc/docs/tutorials/structured_config/intro/) which I haven't allowed, but I believe it allows you to statically type all your parameters and not need the YAML file. I haven't really needed this but super mission-critical services might want this.

BTW, [Lightning-Hydra-Template](https://github.com/ashleve/lightning-hydra-template) which implements best practices also uses hydra?

File: scripts/configs/defaults/augmentations/asymmetric/cifar.yaml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        hue: 0.1
+        prob: 0.8
+      grayscale:
+        enabled: True
+        prob: 0.2
+      gaussian_blur:
+        enabled: True
+        prob: 1.0
+      solarization:
+        enabled: False
```
Comment (DonkeyShot21) [inline scripts/configs/defaults/augmentations/asymmetric/cifar.yaml:21]: isnt it redundant to ave `enabled: False` and `prob: 0.0`? can't we jut remove enabled?

File: solo/methods/barlow_twins.py [hunk 2]
```diff
@@ -17,32 +17,35 @@
...
 from solo.losses.barlow import barlow_loss_func
 from solo.methods.base import BaseMethod
+from solo.utils.misc import omegaconf_select
 
 
 class BarlowTwins(BaseMethod):
-    def __init__(
-        self, proj_hidden_dim: int, proj_output_dim: int, lamb: float, scale_loss: float, **kwargs
-    ):
+    def __init__(self, cfg: omegaconf.DictConfig):
```
Comment (DonkeyShot21) [inline solo/methods/barlow_twins.py:31]: it was nice tho to see the arguments and their type in the init

File: solo/methods/mae.py [hunk 3]
```diff
@@ -164,20 +163,30 @@ def __init__(
...
 
     @staticmethod
-    def add_model_specific_args(parent_parser: argparse.ArgumentParser) -> argparse.ArgumentParser:
-        parent_parser = super(MAE, MAE).add_model_specific_args(parent_parser)
-        parser = parent_parser.add_argument_group("mocov3")
+    def add_and_assert_specific_cfg(cfg: omegaconf.DictConfig) -> omegaconf.DictConfig:
+        """Adds method specific default values/checks for config.
 
-        # decoder
-        parser.add_argument("--decoder_embed_dim", type=int, default=512)
```
Comment (DonkeyShot21) [inline solo/methods/mae.py:172]: missing defaults in mae?

File: solo/methods/barlow_twins.py (hunk 2 above)
Comment (vturrisi) [reply solo/methods/barlow_twins.py:31]: We can have the types in the init, like self.proj_hidden_dim : int = cfg.method_kwargs.proj_hidden_dim

**Comment by turian** [comment]
One more benefit of hydra that I don't think omegaconf has.

Let's say you have two YAML files, one for a super large model and one for a tiny model.

hydra let's you switch configs using just the name of the YAML. Does omegaconf require you to 

(I wanted to take a crack at this but first I'd like to discuss #303 i.e. which is the most common / most covering main file(s) to run on toy data.)

File: scripts/configs/defaults/augmentations/asymmetric/cifar.yaml (hunk 1 above)
Comment (turian) [reply scripts/configs/defaults/augmentations/asymmetric/cifar.yaml]: @DonkeyShot21 hydra (and maybe omegaconf) has this weird constraint that if a config key is not defined, you cannot add it dynamically in code. So perhaps this is why

File: requirements.txt [hunk 4]
```diff
@@ -9,4 +9,5 @@ wandb
 scipy
 timm
 scikit-learn
-omegaconf
\ No newline at end of file
+omegaconf
+hydra
```
Comment (turian) [inline requirements.txt:13]: hydra-core

File: setup.py [hunk 5]
```diff
@@ -57,6 +57,7 @@ def parse_requirements(path):
         "timm",
         "scikit-learn",
         "omegaconf",
+        "hydra",
```
Comment (turian) [inline setup.py:60]: hydra-core

File: scripts/configs/defaults/augmentations/asymmetric/cifar.yaml (hunk 1 above)
Comment (vturrisi) [reply scripts/configs/defaults/augmentations/asymmetric/cifar.yaml]: @turian I think what Enrico meant was to replace `if cfg.solarization.enabled` with `if cfg.solarization.prob > 0`

File: solo/methods/mae.py (hunk 3 above)
Comment (vturrisi) [reply solo/methods/mae.py:172]: Gonna check it later

File: requirements.txt (hunk 4 above)
Comment (turian) [reply requirements.txt:13]: You can also remove omegaconf, since hydra-core pulls it in.

File: setup.py (hunk 5 above)
Comment (turian) [reply setup.py:60]: You can also remove omegaconf, since hydra-core pulls it in.

**Comment by vturrisi** [comment]
I'm thinking about just using hydra to parse the args and merge appropriate default files. Then, we can just call some script like:
python main_pretrain.py --config-path scripts/pretrain/imagenet --config-name barlow {extra parameters to modify} +{extra parameters to add}.

**Comment by turian** [comment]
> I'm thinking about just using hydra to parse the args and merge appropriate default files. Then, we can just call some script like:
> python main_pretrain.py --config-path scripts/pretrain/imagenet --config-name barlow {extra parameters to modify} +{extra parameters to add}.

Agreed

> I'm not a bit fan or relying on hydra for everything (e.g. initializing objs) nor adhering to the default "configs" structure that hydra uses as it seems like a nightmare to manage many settings.

Agreed

Nonetheless, you need to add `hydra-core` to setup.py and requirements.txt. `hydra` is the wrong package.

**Comment by vturrisi** [comment]
Yes, I was in a hurry to commit and put the wrong packaged by mistake. Gonna fix it tomorrow together with moving all config files for pretrain from bash to the new setting. Hopefully I'll also fix the tests soon enough and merge everything this week.

**Comment by turian** [comment]
@vturrisi did CI/CD get disabled?

**Comment by vturrisi** [comment]
@turian it's just failing in the background for some reason. I'm still fixing the tests so it's gonna take a while.

**Pull request merged**